vivisect/impemu/platarch/arm.py

```diff
@@ -1,14 +1,287 @@
+                    # we didn't have a single push in either direction
+                    logger.warn("TOTAL FAILURE TO DETERMINE THUMB MODE")
+                    raise Exception("Neither architecture parsed the first opcode")
+
+                elif armthumb < 0:
+                        self.setFlag(PSR_T_bit, 1)
+                        logger.debug("ArmWorkspaceEmulator: Heuristically Determined funcva is THUMB:  0x%x", funcva)
+                else:
+                    logger.debug("ArmWorkspaceEmulator: Heuristically Determined funcva is ARM:  0x%x", funcva)
+
+
+        self.funcva = funcva
+
+
+    def runFunction(self, funcva, stopva=None, maxhit=None, maxloop=None, tmode=None):
+        """
+        This is a utility function specific to WorkspaceEmulation (and impemu) that
+        will emulate, but only inside the given function.  You may specify a stopva
+        to return once that location is hit.
+        """
+        logger.debug('=== emu.runFunction(0x%x, stopva=%r, maxhit=%r, maxloop=%r, tmode=%r)', funcva, stopva, maxhit, maxloop, tmode)
+        self._prep(funcva, tmode)
+
+        # Let the current (should be base also) path know where we are starting
+        vg_path.setNodeProp(self.curpath, 'bva', funcva)
+
+        hits = {}
+        todo = [(funcva, self.getEmuSnap(), self.path)]
+        vw = self.vw    # Save a dereference many many times
+
+        while len(todo):
+            va, esnap, self.curpath = todo.pop()
+
+            self.setEmuSnap(esnap)
+
+            self.setProgramCounter(va)
+
+            # Check if we are beyond our loop max...
+            if maxloop is not None:
+                lcount = vg_path.getPathLoopCount(self.curpath, 'bva', va)
+                if lcount > maxloop:
+                    continue
+
+            while True:
+
+                starteip = self.getProgramCounter()
+
+                if not vw.isValidPointer(starteip):
+                    break
+
+                if starteip == stopva:
+                    return
+
+                # Check straight hit count...
+                if maxhit != None:
+                    h = hits.get(starteip, 0)
+                    h += 1
+                    if h > maxhit:
+                        break
+                    hits[starteip] = h
+
+                # If we ran out of path (branches that went
+                # somewhere that we couldn't follow?
+                if self.curpath == None:
+                    break
+
+                try:
+
+                    tmode = self.getFlag(PSR_T_bit)
+                    # FIXME unify with stepi code...
+                    op = self.parseOpcode(starteip | tmode)
+
+                    self.op = op
+                    if self.emumon:
+                        try:
+                            self.emumon.prehook(self, op, starteip)
+                        except Exception, e:
+                            if not self.getMeta('silent'):
```

Comment:
This is when an emulator is being used in "maybe" code...it's extraordinarily noisy and confusing when trying to track down potential bugs.

Reply:
it's something for analysis not necessarily already in use in the Vivisect codebase.  however, this would be great to have in the isCode() analysis, since we're strapping some random bytes into the emulator and trying to emulate them.  that can be terribly noisy.  perhaps we should wrap this (or something like it) into the main WorkspaceEmulator.

Reply:
actually, made this more pertinent now.  moved it to the main workspace emulator def and set it in isProbablyCode()